### PR TITLE
[pilot] add option for `uses_non_exempt_encryption` and wait for processing again when being set

### LINKED ISSUE #14033: Unable to distribute to testflight builds
### Question Checklist

- [X] Updated fastlane to the latest version
- [X] I read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [X] I read [docs.fastlane.tools](https://docs.fastlane.tools)
- [X] I searched for [existing GitHub issues](https://github.com/fastlane/fastlane/issues)

### Question Subject
Error when distributing testflight builds via pilot. 

#### Question Description
We started seeing these errors in our CI pipe line a couple weeks ago that says we are unable to update the change log. But seems like there is something else going on here. 
As we haven't made any changes to our fastlane scripts, anyone else having issues like this starting a couple weeks ago?

Here's are verbose log:

```
ERROR [2019-01-07 17:11:12.56]: fastlane finished with errors
Traceback (most recent call last):
	24: from /Users/chwon/.rvm/gems/ruby-2.5.1/bin/ruby_executable_hooks:24:in `<main>'
	23: from /Users/chwon/.rvm/gems/ruby-2.5.1/bin/ruby_executable_hooks:24:in `eval'
	22: from /Users/chwon/.rvm/gems/ruby-2.5.1/bin/fastlane:23:in `<main>'
	21: from /Users/chwon/.rvm/gems/ruby-2.5.1/bin/fastlane:23:in `load'
	20: from /Users/chwon/.rvm/gems/ruby-2.5.1@global/gems/fastlane-2.113.0/bin/fastlane:23:in `<top (required)>'
	19: from /Users/chwon/.rvm/rubies/ruby-2.5.1/lib/ruby/gems/2.5.0/gems/fastlane-2.113.0/fastlane/lib/fastlane/cli_tools_distributor.rb:108:in `take_off'
	18: from /Users/chwon/.rvm/rubies/ruby-2.5.1/lib/ruby/gems/2.5.0/gems/fastlane-2.113.0/fastlane/lib/fastlane/commands_generator.rb:41:in `start'
	17: from /Users/chwon/.rvm/rubies/ruby-2.5.1/lib/ruby/gems/2.5.0/gems/fastlane-2.113.0/fastlane/lib/fastlane/commands_generator.rb:333:in `run'
	16: from /Users/chwon/.rvm/gems/ruby-2.5.1/gems/commander-fastlane-4.4.6/lib/commander/delegates.rb:15:in `run!'
	15: from /Users/chwon/.rvm/rubies/ruby-2.5.1/lib/ruby/gems/2.5.0/gems/fastlane-2.113.0/fastlane_core/lib/fastlane_core/ui/fastlane_runner.rb:76:in `run!'
	14: from /Users/chwon/.rvm/gems/ruby-2.5.1/gems/commander-fastlane-4.4.6/lib/commander/runner.rb:476:in `run_active_command'
	13: from /Users/chwon/.rvm/gems/ruby-2.5.1/gems/commander-fastlane-4.4.6/lib/commander/command.rb:153:in `run'
	12: from /Users/chwon/.rvm/gems/ruby-2.5.1/gems/commander-fastlane-4.4.6/lib/commander/command.rb:178:in `call'
	11: from /Users/chwon/.rvm/rubies/ruby-2.5.1/lib/ruby/gems/2.5.0/gems/fastlane-2.113.0/fastlane/lib/fastlane/commands_generator.rb:108:in `block (2 levels) in run'
	10: from /Users/chwon/.rvm/rubies/ruby-2.5.1/lib/ruby/gems/2.5.0/gems/fastlane-2.113.0/fastlane/lib/fastlane/command_line_handler.rb:36:in `handle'
	 9: from /Users/chwon/.rvm/rubies/ruby-2.5.1/lib/ruby/gems/2.5.0/gems/fastlane-2.113.0/fastlane/lib/fastlane/lane_manager.rb:56:in `cruise_lane'
	 8: from /Users/chwon/.rvm/rubies/ruby-2.5.1/lib/ruby/gems/2.5.0/gems/fastlane-2.113.0/fastlane/lib/fastlane/runner.rb:45:in `execute'
	 7: from /Users/chwon/.rvm/rubies/ruby-2.5.1/lib/ruby/gems/2.5.0/gems/fastlane-2.113.0/fastlane/lib/fastlane/runner.rb:45:in `chdir'
	 6: from /Users/chwon/.rvm/rubies/ruby-2.5.1/lib/ruby/gems/2.5.0/gems/fastlane-2.113.0/fastlane/lib/fastlane/runner.rb:49:in `block in execute'
	 5: from /Users/chwon/.rvm/rubies/ruby-2.5.1/lib/ruby/gems/2.5.0/gems/fastlane-2.113.0/fastlane/lib/fastlane/lane.rb:33:in `call'
	 4: from ../../Common/fastlane/Fastfile:382:in `block (2 levels) in parsing_binding'
	 3: from /Users/chwon/.rvm/rubies/ruby-2.5.1/lib/ruby/gems/2.5.0/gems/fastlane-2.113.0/pilot/lib/pilot/build_manager.rb:88:in `distribute'
	 2: from /Users/chwon/.rvm/rubies/ruby-2.5.1/lib/ruby/gems/2.5.0/gems/fastlane-2.113.0/pilot/lib/pilot/build_manager.rb:92:in `rescue in distribute'
	 1: from /Users/chwon/.rvm/rubies/ruby-2.5.1/lib/ruby/gems/2.5.0/gems/fastlane-2.113.0/fastlane_core/lib/fastlane_core/ui/ui.rb:17:in `method_missing'
/Users/chwon/.rvm/rubies/ruby-2.5.1/lib/ruby/gems/2.5.0/gems/fastlane-2.113.0/fastlane_core/lib/fastlane_core/ui/interface.rb:141:in `user_error!': [!] Could not set changelog:


<html> (FastlaneCore::Interface::FastlaneError)
<head>
<meta http-equiv="Content-Type" content="text/html;charset=ISO-8859-1"/>
<title>Error 409 </title>
</head>
<body>
<h2>HTTP ERROR: 409</h2>
<p>Problem accessing /v2/providers/302756/apps/1382945144/builds/37665298. Reason:
<pre>    ERROR-1: Enter a demo account user name, or deselect "A demo account is required to use all the features of the app."; ERROR-1: Enter demo account password, or deselect "A demo account is required to use all the features of the app."</pre></p>
<hr /><i><small>Powered by Jetty://</small></i>
</body>
</html>
```

Other jobs failed with this error: The build is not in state for internal testing. - Build is not in internal testing state.

```
/usr/local/lib/ruby/gems/2.5.0/gems/fastlane-2.112.0/spaceship/lib/spaceship/connect_api/client.rb:84:in `handle_response': [!] The build is not in state for internal testing. - Build is not in internal testing state. (Spaceship::UnexpectedResponse)
	from /usr/local/lib/ruby/gems/2.5.0/gems/fastlane-2.112.0/spaceship/lib/spaceship/connect_api/client.rb:66:in `post_for_testflight_review'
	from /usr/local/lib/ruby/gems/2.5.0/gems/fastlane-2.112.0/spaceship/lib/spaceship/test_flight/build.rb:223:in `submit_for_testflight_review!'
	from /usr/local/lib/ruby/gems/2.5.0/gems/fastlane-2.112.0/pilot/lib/pilot/build_manager.rb:183:in `distribute_build'
	from /usr/local/lib/ruby/gems/2.5.0/gems/fastlane-2.112.0/pilot/lib/pilot/build_manager.rb:108:in `distribute'
	from ../../Common/fastlane/Fastfile:382:in `block (2 levels) in parsing_binding'
	from /usr/local/lib/ruby/gems/2.5.0/gems/fastlane-2.112.0/fastlane/lib/fastlane/lane.rb:33:in `call'
	from /usr/local/lib/ruby/gems/2.5.0/gems/fastlane-2.112.0/fastlane/lib/fastlane/runner.rb:49:in `block in execute'
	from /usr/local/lib/ruby/gems/2.5.0/gems/fastlane-2.112.0/fastlane/lib/fastlane/runner.rb:45:in `chdir'
	from /usr/local/lib/ruby/gems/2.5.0/gems/fastlane-2.112.0/fastlane/lib/fastlane/runner.rb:45:in `execute'
	from /usr/local/lib/ruby/gems/2.5.0/gems/fastlane-2.112.0/fastlane/lib/fastlane/lane_manager.rb:56:in `cruise_lane'
	from /usr/local/lib/ruby/gems/2.5.0/gems/fastlane-2.112.0/fastlane/lib/fastlane/command_line_handler.rb:36:in `handle'
	from /usr/local/lib/ruby/gems/2.5.0/gems/fastlane-2.112.0/fastlane/lib/fastlane/commands_generator.rb:108:in `block (2 levels) in run'
	from /usr/local/lib/ruby/gems/2.5.0/gems/commander-fastlane-4.4.6/lib/commander/command.rb:178:in `call'
	from /usr/local/lib/ruby/gems/2.5.0/gems/commander-fastlane-4.4.6/lib/commander/command.rb:153:in `run'
	from /usr/local/lib/ruby/gems/2.5.0/gems/commander-fastlane-4.4.6/lib/commander/runner.rb:476:in `run_active_command'
	from /usr/local/lib/ruby/gems/2.5.0/gems/fastlane-2.112.0/fastlane_core/lib/fastlane_core/ui/fastlane_runner.rb:76:in `run!'
	from /usr/local/lib/ruby/gems/2.5.0/gems/commander-fastlane-4.4.6/lib/commander/delegates.rb:15:in `run!'
	from /usr/local/lib/ruby/gems/2.5.0/gems/fastlane-2.112.0/fastlane/lib/fastlane/commands_generator.rb:333:in `run'
	from /usr/local/lib/ruby/gems/2.5.0/gems/fastlane-2.112.0/fastlane/lib/fastlane/commands_generator.rb:41:in `start'
	from /usr/local/lib/ruby/gems/2.5.0/gems/fastlane-2.112.0/fastlane/lib/fastlane/cli_tools_distributor.rb:108:in `take_off'
	from /usr/local/lib/ruby/gems/2.5.0/gems/fastlane-2.112.0/bin/fastlane:23:in `<top (required)>'
	from /usr/local/bin/fastlane:23:in `load'
	from /usr/local/bin/fastlane:23:in `<main>'
```

### PR DESCRIPTION
**Pull request by joshdholtz**
Fixes #14033

### Motivation and Context
```sh
2019-03-05T01:14:00.4929920Z Spaceship::UnexpectedResponse: �[31m[!] The build is not in state for internal testing. - Build is not in internal testing state.�[0m
2019-03-05T01:14:00.4930960Z   /drop/vendor/bundle/ruby/2.3.0/gems/fastlane-2.117.1/spaceship/lib/spaceship/connect_api/client.rb:274:in `handle_response'
```
Taken from https://github.com/fastlane/fastlane/issues/14033#issuecomment-469615095

Soooo, the issue that caused the `The build is not in state for internal testing. - Build is not in internal testing state.` ended up being that if you **do not** provide the `ITSAppUsesNonExemptEncryption` key in your `Info.plist` and instead rely on `pilot` setting it for you, the build goes **back** into the `PROCESSING` state for about a minute or a few 😬

### Description
So the best solution is to add the `ITSAppUsesNonExemptEncryption` key to your `Info.plist` to make the _fastlane_ part take up less time so it only needs to process once but... I'm adding a fix where we will be waiting for the build to finish processing again between the setting of the export compliance and the actual submission.

I also added a new `uses_non_exempt_encryption` option since we shouldn't hardcode `false` for users and some users might need to set `true`

#### Output without `ITSAppUsesNonExemptEncryption`

<details>

```sh
[21:29:26]: -------------------
[21:29:26]: --- Step: pilot ---
[21:29:26]: -------------------
[21:29:26]: Login to App Store Connect (josh+fastlane@rokkincat.com)
[21:29:29]: Login successful
[21:29:30]: Ready to upload new build to TestFlight (App: 12341234)...
[21:29:31]: Going to upload updated app to App Store Connect
[21:29:31]: This might take a few minutes. Please don't interrupt the script.
[21:31:51]: iTunes Transporter successfully finished its job
[21:31:51]: ------------------------------------------------------------------------------------------------------------------
[21:31:51]: --- Successfully uploaded package to App Store Connect. It might take a few minutes until it's visible online. ---
[21:31:51]: ------------------------------------------------------------------------------------------------------------------
[21:31:51]: Successfully uploaded the new binary to App Store Connect
[21:31:51]: If you want to skip waiting for the processing to be finished, use the `skip_waiting_for_build_processing` option
[21:31:55]: Build doesn't show up in the build list anymore, waiting for it to appear again (check your email for processing issues if this continues)
[21:32:29]: Waiting for App Store Connect to finish processing the new build (1.0 - 157)
[21:33:03]: Waiting for App Store Connect to finish processing the new build (1.0 - 157)
[21:33:37]: Waiting for App Store Connect to finish processing the new build (1.0 - 157)
[21:34:11]: Waiting for App Store Connect to finish processing the new build (1.0 - 157)
[21:34:44]: Waiting for App Store Connect to finish processing the new build (1.0 - 157)
[21:35:17]: Waiting for App Store Connect to finish processing the new build (1.0 - 157)
[21:35:50]: Waiting for App Store Connect to finish processing the new build (1.0 - 157)
[21:36:23]: Waiting for App Store Connect to finish processing the new build (1.0 - 157)
[21:36:57]: Waiting for App Store Connect to finish processing the new build (1.0 - 157)
[21:37:30]: Waiting for App Store Connect to finish processing the new build (1.0 - 157)
[21:38:05]: Waiting for App Store Connect to finish processing the new build (1.0 - 157)
[21:38:38]: Waiting for App Store Connect to finish processing the new build (1.0 - 157)
[21:39:13]: Successfully finished processing the build 1.0 - 157
[21:39:19]: Successfully set the changelog for build
[21:39:21]: Distributing new build to testers: 1.0 - 157
[21:39:23]: Export comlpiance has been set to 'false'. Need to wait for build to finishing processing again....
[21:39:23]: Set 'ITSAppUsesNonExemptEncryption' in the 'Info.plist' to skip this step and speed up the submission
[21:39:27]: Build doesn't show up in the build list anymore, waiting for it to appear again (check your email for processing issues if this continues)
[21:40:01]: Build doesn't show up in the build list anymore, waiting for it to appear again (check your email for processing issues if this continues)
[21:40:35]: Build doesn't show up in the build list anymore, waiting for it to appear again (check your email for processing issues if this continues)
[21:41:08]: Build doesn't show up in the build list anymore, waiting for it to appear again (check your email for processing issues if this continues)
[21:41:56]: Build doesn't show up in the build list anymore, waiting for it to appear again (check your email for processing issues if this continues)
[21:42:31]: Build doesn't show up in the build list anymore, waiting for it to appear again (check your email for processing issues if this continues)
[21:43:05]: Build doesn't show up in the build list anymore, waiting for it to appear again (check your email for processing issues if this continues)
[21:43:38]: Build doesn't show up in the build list anymore, waiting for it to appear again (check your email for processing issues if this continues)
[21:44:12]: Successfully finished processing the build 1.0 - 157
[21:44:18]: Successfully distributed build to External testers 🚀
+------+------------------------+-------------+
|              fastlane summary               |
+------+------------------------+-------------+
| Step | Action                 | Time (in s) |
+------+------------------------+-------------+
| 1    | increment_build_number | 1           |
| 2    | gym                    | 52          |
| 3    | pilot                  | 891         |
+------+------------------------+-------------+
[21:44:18]: fastlane.tools just saved you 16 minutes! 🎉
```

</details>

#### Output with `ITSAppUsesNonExemptEncryption`

<details>

```sh
[21:14:50]: -------------------
[21:14:50]: --- Step: pilot ---
[21:14:50]: -------------------
[21:14:50]: Login to App Store Connect (josh+fastlane@rokkincat.com)
[21:14:52]: Login successful
[21:14:53]: Ready to upload new build to TestFlight (App: 12341234)...
[21:14:53]: Going to upload updated app to App Store Connect
[21:14:53]: This might take a few minutes. Please don't interrupt the script.
[21:16:44]: iTunes Transporter successfully finished its job
[21:16:44]: ------------------------------------------------------------------------------------------------------------------
[21:16:44]: --- Successfully uploaded package to App Store Connect. It might take a few minutes until it's visible online. ---
[21:16:44]: ------------------------------------------------------------------------------------------------------------------
[21:16:44]: Successfully uploaded the new binary to App Store Connect
[21:16:44]: If you want to skip waiting for the processing to be finished, use the `skip_waiting_for_build_processing` option
[21:16:48]: Build doesn't show up in the build list anymore, waiting for it to appear again (check your email for processing issues if this continues)
[21:17:22]: Waiting for App Store Connect to finish processing the new build (1.0 - 156)
[21:17:55]: Waiting for App Store Connect to finish processing the new build (1.0 - 156)
[21:18:28]: Waiting for App Store Connect to finish processing the new build (1.0 - 156)
[21:19:02]: Waiting for App Store Connect to finish processing the new build (1.0 - 156)
[21:19:34]: Waiting for App Store Connect to finish processing the new build (1.0 - 156)
[21:20:18]: Waiting for App Store Connect to finish processing the new build (1.0 - 156)
[21:20:52]: Waiting for App Store Connect to finish processing the new build (1.0 - 156)
[21:21:27]: Waiting for App Store Connect to finish processing the new build (1.0 - 156)
[21:22:09]: Waiting for App Store Connect to finish processing the new build (1.0 - 156)
[21:22:42]: Waiting for App Store Connect to finish processing the new build (1.0 - 156)
[21:23:15]: Waiting for App Store Connect to finish processing the new build (1.0 - 156)
[21:23:48]: Waiting for App Store Connect to finish processing the new build (1.0 - 156)
[21:24:21]: Waiting for App Store Connect to finish processing the new build (1.0 - 156)
[21:24:55]: Build doesn't show up in the build list anymore, waiting for it to appear again (check your email for processing issues if this continues)
[21:25:28]: Build doesn't show up in the build list anymore, waiting for it to appear again (check your email for processing issues if this continues)
[21:26:02]: Build doesn't show up in the build list anymore, waiting for it to appear again (check your email for processing issues if this continues)
[21:26:36]: Build doesn't show up in the build list anymore, waiting for it to appear again (check your email for processing issues if this continues)
[21:27:09]: Build doesn't show up in the build list anymore, waiting for it to appear again (check your email for processing issues if this continues)
[21:27:43]: Successfully finished processing the build 1.0 - 156
[21:27:48]: Successfully set the changelog for build
[21:27:50]: Distributing new build to testers: 1.0 - 156
[21:27:57]: Successfully distributed build to External testers 🚀
+------+------------------------+-------------+
|              fastlane summary               |
+------+------------------------+-------------+
| Step | Action                 | Time (in s) |
+------+------------------------+-------------+
| 1    | increment_build_number | 1           |
| 2    | gym                    | 39          |
| 3    | pilot                  | 787         |
+------+------------------------+-------------+
[21:27:57]: fastlane.tools just saved you 14 minutes! 🎉
```

</details>

### CODE REVIEW

File: pilot/lib/pilot/build_manager.rb [hunk 1]
```diff
@@ -241,16 +247,27 @@ def distribute_build(uploaded_build, options)
...
       end
 
       true
     end
 
+    def set_export_compliance_if_needed(uploaded_build, options)
+      build = uploaded_build.find_app_store_connect_build
+      build_attributes = build["attributes"] || {}
+      if build_attributes["usesNonExemptEncryption"].nil?
+        uses_non_exempt_encryption = options[:uses_non_exempt_encryption]
```
Comment (taquitos) [inline pilot/lib/pilot/build_manager.rb]: shouldn't we always set `uses_non_exempt_encryption` if the option `options[:uses_non_exempt_encryption]` provides it? That means the developer wants to override the plist, right?

File: spaceship/lib/spaceship/connect_api/client.rb [hunk 2]
```diff
@@ -230,6 +230,27 @@ def get_builds(filter: {}, includes: "buildBetaDetail,betaBuildMetrics", limit:
         handle_response(response)
       end
 
+      def patch_builds(build_id: nil, attributes: {})
```
Comment (taquitos) [inline spaceship/lib/spaceship/connect_api/client.rb]: Maybe add more of a description of when/why this function would be called

File: pilot/lib/pilot/build_manager.rb (hunk 1 above)
Comment (taquitos) [reply pilot/lib/pilot/build_manager.rb]: I guess since we use a default, we can't really tell if the plist if being overridden by the developer.

Comment (joshdholtz) [reply pilot/lib/pilot/build_manager.rb]: @taquitos I currently have `uses_non_exempt_encryption` set to `default_value: false` so I don't want to override it if the plist is set to `true`. I would assume the plist values takes precedence over this. If the plist values is set then `usesNonExemptEncryption` will have a value and we skip it.

So if we always want to set this then maybe I should remove the `default_value: false` and make `optional: true`? But if that is the case then we might run into an issue where plist isn't set and the option isn't set so then we run into this same state 😬

Comment (taquitos) [reply pilot/lib/pilot/build_manager.rb]: Yeah, let's leave it as is. I like your approach.

**Comment by joshdholtz** [comment]
I got a 👍 via Slack so going to merge this 🙃

**Review by taquitos — CHANGES_REQUESTED** [review]
🎈🐐

**Review by taquitos — APPROVED** [review]

**Pull request merged**